### ISSUE TITLE
TF Memory Leak

### ISSUE THREAD

**Issue by jonah-chen** [jonah-chen/alphazero-guerzhoy#4]
Memory Leak issue.
![image](https://user-images.githubusercontent.com/71615343/102152180-55dc1500-3e29-11eb-90c7-4b1b9b95d0e0.png)

Upon consultation with Prof. Guerzhoy, he suggest it is likely caused to one of the tensorflow functions. Will attempt to debug with memory-profiler library (from https://pypi.org/project/memory-profiler/) to pinpoint the exact issue before trying to resolve.

**Comment by jonah-chen** [comment]
Executing python's memory profiler (https://pypi.org/project/memory-profiler/) for about 10 hours, the results are shown here. If one can help with how this is interpreted, that will be really appreciated as I'm having a hard time seeing what is causing the issue and how to resolve it. 

`Line #    Mem usage    Increment  Occurences   Line Contents
============================================================
   159 28167.809 MiB 2386942.426 MiB         270   @profile
   160                                         def optimized_search(model, boards, players, it=512, roots=None):
   161                                             """Perform Monte-Carlo Tree Search on a batch of boards with it iterations. 
   162                                             Return a list of Node objects, whose children is selected with with a given probability distribution for the next move.
   163                                             More optimized version since it takes a longer per board to predict an array with less boards 
   164                                             """
   165 28167.809 MiB    0.258 MiB         270       games = len(boards)
   166                                         
   167                                             # Store all the root nodes that need to be evaluated
   168 28167.809 MiB    0.000 MiB         270       if roots is None:
   169 27425.883 MiB    0.000 MiB         565           roots = [Node(0, player) for player in players]
   170                                                 # Calculate the policies the root nodes
   171                                                 # Note the values of the root nodes are irrelevent
   172 27425.883 MiB  482.648 MiB        1125           policies, _ = model(np.array([boards[i] if players[i] == 1 else np.flip(
   173 27425.883 MiB    0.000 MiB         560               boards[i], axis=2) for i in range(games)]), training=False)
   174                                                 
   175 27425.883 MiB    0.000 MiB           5           policies = policies.numpy()
   176                                         
   177                                                 # Expand all root nodes
   178 27425.883 MiB    0.000 MiB         555           for i in range(games):
   179 27425.883 MiB    3.426 MiB         550               roots[i].expand(policies[i], boards[i])
   180                                         
   181 28167.809 MiB    0.000 MiB         270       start = perf_counter()
   182                                             # Performs the search
   183 28177.621 MiB -6276.398 MiB      138309       for _ in range(it):
   184 28177.621 MiB -6350.941 MiB      138040           if _ % 32 == 31:
   185 28173.559 MiB -146.840 MiB        4313               end = perf_counter()
   186 28173.559 MiB -146.840 MiB        4313               print(f'iteration:{_+1} of {it} time:{(end-start):.2f}s')
   187 28173.559 MiB -146.840 MiB        4313               start = perf_counter()
   188                                         
   189 28177.621 MiB -3251.496 MiB      138040           search_boards = deepcopy(boards)
   190 28177.621 MiB -5653.965 MiB      138040           nodes = copy(roots)
   191 28177.621 MiB -5654.840 MiB      138040           search_players = copy(players)
   192                                                 # numpy transpose may or may not be faster
   193 28177.621 MiB -1140790.711 MiB    11329688           paths = [[node] for node in nodes]
   194                                         
   195                                                 # For all games, play maximizing Q+U until a leaf node is reached.
   196 28177.621 MiB -1129493.520 MiB    11053605           for i in range(games):
   197 28177.621 MiB -6045122.500 MiB    70859925               while nodes[i].expanded:
   198 28177.621 MiB -4921283.703 MiB    59944360                   action, nodes[i] = nodes[i].select()
   199 28177.621 MiB -4921281.891 MiB    59944359                   paths[i].append(nodes[i])
   200 28177.621 MiB -9842566.352 MiB   119888718                   move_on_board(search_boards[i],
   201 28177.621 MiB -4921283.535 MiB    59944359                                 action, player=search_players[i])
   202 28177.621 MiB -4921283.703 MiB    59944359                   search_players[i] = search_players[i] % 2 + 1
   203                                         
   204                                                 # Flips the boards of player 2 so that it can be evaluated by neural network
   205                                         
   206 28177.621 MiB -1129493.520 MiB    11053557           for i in range(games):
   207 28177.621 MiB -1123838.797 MiB    10915518               if search_players[i] != 1:
   208 28177.621 MiB -313376.695 MiB     4398969                   search_boards[i] = np.flip(search_boards[i], axis=2)
   209                                         
   210                                                 # Evaulate: Determine the policies and values as well and the end state of the leaf nodes.
   211                                         
   212                                                 # Performs at random one of eight symmetries of the board.
   213 28177.621 MiB -5654.676 MiB      138039           rotation = random.randint(0, 3)
   214 28177.621 MiB -5654.723 MiB      138039           reflection = random.randint(0, 1)
   215                                         
   216                                                 # Calculate the policies and values using the NN and reverses the transformation.
   217                                         
   218 28177.621 MiB -5654.723 MiB      138039           if rotation:
   219 28177.621 MiB -4050.000 MiB      103453               if reflection:
   220 28177.621 MiB -391.332 MiB       51935                   policies, values = model(np.rot90(np.flip(search_boards, axis=2), rotation, (1, 2)), training=False)
   221 28177.621 MiB -1861.430 MiB       51935                   policies, values = policies.numpy(), values.numpy().flatten()
   222 28177.621 MiB -5595.152 MiB      155805                   policies = np.flip(
   223 28177.621 MiB -3730.102 MiB      103870                       np.rot90(policies.reshape(-1, 8, 8,),
   224 28177.621 MiB -5595.152 MiB      155805                                rotation, (2, 1)), axis=2).reshape(-1, 64,)
   225                                                     else:
   226 28176.180 MiB  -76.746 MiB       51518                   policies, values = model(np.rot90(search_boards, rotation, (1, 2)), training=False)
   227 28176.180 MiB -1860.141 MiB       51518                   policies, values = policies.numpy(), values.numpy().flatten()
   228 28176.180 MiB -5586.277 MiB      154554                   policies = np.rot90(policies.reshape(-1, 8, 8,),
   229 28176.180 MiB -3724.281 MiB      103036                                       rotation, (2, 1)).reshape(-1, 64,)
   230 28176.898 MiB -1604.723 MiB       34586           elif reflection:
   231 28174.996 MiB   57.500 MiB       17283               policies, values = model(np.flip(search_boards, axis=2), training=False)
   232 28174.996 MiB -541.188 MiB       17283               policies, values = policies.numpy(), values.numpy().flatten()
   233 28174.996 MiB -1625.672 MiB       51849               policies = np.flip(policies.reshape(-1, 8, 8,),
   234 28174.996 MiB -1083.781 MiB       34566                                  axis=2).reshape(-1, 64,)
   235                                                 else:
   236 28176.898 MiB -356.453 MiB       17303               policies, values = model(search_boards, training=False)
   237 28177.145 MiB -155.480 MiB       17303               policies, values = policies.numpy(), values.numpy().flatten()
   238                                         
   239                                                 
   240                                         
   241 28177.621 MiB -1154938.527 MiB    11329635           states = [is_win(search_board) for search_board in search_boards]
   242                                         
   243                                                 # Correct the evaluation for games with end states. -1.0 for loss, 0.0 for draw, 1.0 for win.
   244 28177.621 MiB -1222503.086 MiB    11053557           for i in range(games):
   245 28177.621 MiB -1217839.051 MiB    10915518               if states[i] == 1:
   246                                                         values[i] = 1.0
   247 28177.621 MiB -1217839.051 MiB    10915518               elif states[i] == 2:
   248 28177.379 MiB -1892.211 MiB      610699                   values[i] = -1.0
   249 28177.621 MiB -1215946.586 MiB    10304819               elif states[i] == 3:
   250 27065.820 MiB -291.562 MiB      594052                   values[i] == 0.0
   251                                                     else:
   252 28177.621 MiB -1202427.645 MiB     9710767                   nodes[i].expand(policies[i], search_boards[i])
   253                                         
   254 28177.621 MiB -1218909.969 MiB    11053557           for i in range(games):
   255 28177.621 MiB -6509123.703 MiB    70859501               for bnode in paths[i][:0:-1]:
   256 28177.621 MiB -5296491.691 MiB    59943983                   bnode.N += 1
   257 28177.621 MiB -5296491.719 MiB    59943983                   if bnode.player == search_players[i]:
   258 28177.621 MiB -2819790.793 MiB    32733878                       bnode.W += values[i]
   259                                                         else:
   260 28177.621 MiB -2476684.414 MiB    27210105                       bnode.W -= values[i]
   261 28167.809 MiB    0.000 MiB         269       return roots


Filename: /home/hina/Documents/CODE/alphazero-guerzhoy/play.py

Line #    Mem usage    Increment  Occurences   Line Contents
============================================================
    18 20195.551 MiB 21307.344 MiB           2   @profile
    19                                         def self_play(model, games=128, game_iter=64, search_iter=512):
    20 20195.551 MiB    0.000 MiB           2       boards = np.zeros((games, 8, 8, 2,), dtype="float32")
    21 20195.551 MiB    0.000 MiB           2       players = [1]*games
    22 20195.551 MiB    0.000 MiB           2       inputs = None
    23                                         
    24 20195.551 MiB    0.000 MiB           2       s = []
    25 20195.551 MiB    0.000 MiB           2       pie = []
    26 20195.551 MiB    0.000 MiB           2       z = []
    27                                         
    28                                             # These are the parameters to train the network to gained by MCTS process
    29                                             # The elements are accessed as game_boards[game#][turn#]
    30 20195.551 MiB    0.000 MiB         406       game_boards = [[] for _ in range(games)]
    31 20195.551 MiB    0.000 MiB         406       mcts_policies = [[] for _ in range(games)]
    32                                         
    33 27066.074 MiB    0.000 MiB         130       for turns in range(game_iter):
    34 27020.469 MiB    0.000 MiB         256           print(
    35 27020.469 MiB    0.000 MiB         128               f"------------------------------------------------------------\nTurn {turns+1} of {game_iter}. Ended: {games - len(game_boards)} of {games}. Cumulated: {int(perf_counter() - true_start)}s")
    36 27020.469 MiB    0.000 MiB         128           if len(game_boards) == 0:
    37                                                     return s, pie, z
    38 27065.820 MiB 2364879.461 MiB         256           results = optimized_search(
    39 27020.469 MiB    0.000 MiB         128               model, boards, players, roots=inputs, it=search_iter)
    40 27065.820 MiB    0.000 MiB         128           inputs = []
    41 27065.820 MiB    0.000 MiB         128           games_ended = 0
    42 27066.074 MiB    0.000 MiB       14597           for j in range(len(results)):
    43 27066.074 MiB    0.000 MiB       14469               i = j - games_ended
    44                                         
    45                                                     # Save the results of the MCTS to train NN
    46 27066.074 MiB    5.668 MiB       14469               act, dist = results[i].play()
    47 27066.074 MiB    0.000 MiB       28938               game_boards[i].append(
    48 27066.074 MiB    1.598 MiB       14469                   deepcopy(boards[i] if players[i] == 1 else np.flip(boards[i], axis=2)))
    49 27066.074 MiB    0.000 MiB       14469               mcts_policies[i].append(dist)
    50                                         
    51                                                     # Make Move
    52 27066.074 MiB    0.000 MiB       14469               move_on_board(boards[i], act, player=players[i])
    53                                         
    54                                                     # When game ends, save the data of the game.
    55 27066.074 MiB    0.098 MiB       14469               state = is_win(boards[i])
    56 27066.074 MiB    0.000 MiB       14469               if state:
    57 27066.074 MiB    0.000 MiB         400                   s.append(game_boards.pop(i))
    58 27066.074 MiB    0.000 MiB         400                   pie.append(mcts_policies.pop(i))
    59 27066.074 MiB    0.000 MiB         400                   if state == 1:
    60 26569.398 MiB    0.000 MiB        5530                       z.append([1 - 2 * (k % 2) for k in range(turns+1)])
    61 27066.074 MiB    0.000 MiB         193                   elif state == 2:
    62 26611.668 MiB    0.000 MiB        2970                       z.append([2 * (k % 2) - 1 for k in range(turns+1)])
    63 27066.074 MiB    0.000 MiB         107                   elif state == 3:
    64 27066.074 MiB    0.000 MiB         107                       z.append([0]*(turns+1))
    65 27066.074 MiB    0.500 MiB         400                   boards = np.delete(boards, i, axis=0)
    66 27066.074 MiB    0.000 MiB         400                   players.pop()
    67 27066.074 MiB    0.000 MiB         400                   del results[i]
    68                                         
    69 27066.074 MiB    0.000 MiB         400                   games_ended += 1
    70                                                     else:
    71                                                         # When game doesn't end. Player changes and the new state is appended to be evaluated on the next tern.
    72 27020.469 MiB    0.000 MiB       14069                   inputs.append(results[i].children[act])
    73 27020.469 MiB    0.000 MiB       14069                   players[i] = players[i] % 2 + 1
    74                                         
    75 27066.074 MiB    0.000 MiB           2       return s, pie, z
`

**Comment by MAK13789** [comment]
![image](https://user-images.githubusercontent.com/24789906/102699616-083e1e80-4214-11eb-9d51-90be097b3013.png)

**Comment by jonah-chen** [comment]
A memory leak is closed on `static PyObject* is_win(PyObject* self, PyObject* args)` in the C-Python module by using `int PyArray_Free(PyObject* op, void* ptr)` function instead of the default `free` function. Before running speedtest.py will incur a large amount of memory usage ~2-3GB/s. With this fix, that no longer happens. However, not sure if there are other memory leaks thus a test will be run to see if this resolves the memory leak issue.
